src/v/io/interval_map.h

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include <absl/container/btree_map.h>
+
+#include <concepts>
+#include <utility>
+
+namespace experimental::io {
+
+/**
+ * A container that maps intervals to values.
+ *
+ * The interval_map holds non-overlapping, non-empty, open intervals, and
+ * associates each interval with a given value. For example:
+ *
+ *     [0000, 4096) -> Page0
+ *     [4096, 8192) -> Page1
+ */
+template<std::integral T, typename V>
+class interval_map {
+    using map_type = absl::btree_map<T, std::pair<T, V>>;
```

Comment:
I wonder if this will be more readable for both callers and the implementation (because nobody has to figure out what `it->second.first` means) if we do the following:

```cpp
struct entry {
  T start;
  T length;
  V value;
  
  bool operator<(const entry& other) const {
    return first < other.start;
  }
}
using map_type = absl::btree_map<T, std::pair<T, V>>;
```

Reply:
Are you thinking for the user of this class? If so, I agree, though I might be tempted to make the key be a std::pair<int,int> so that the value can still be separate.

For what its worth, this annoying `first.second.second` business doesn't leak very far since I've wrapped this for the actual page set to provide a set / page oriented interface:

```c++
class page_set {
    using map_type = io::interval_map<uint64_t, seastar::lw_shared_ptr<page>>;

public:
    class const_iterator
      : public boost::iterator_adaptor<
          const_iterator,
          map_type::const_iterator,
          const seastar::lw_shared_ptr<page>,
          boost::iterators::forward_traversal_tag> {
    private:
        friend class boost::iterator_core_access;
        friend class page_set;
        explicit const_iterator(map_type::const_iterator);
        reference dereference() const;
    };

    const_iterator find(uint64_t offset) const;
    void insert(seastar::lw_shared_ptr<page> page);
...
```

Reply:
I added a commit which switch over the implementation to use an interval structure for the key. i'm not sure it's significantly easier to read, but it is definitely better.

Reply:
I like it. Thanks